### ISSUE TITLE
Support for more client apps

### ISSUE THREAD

**Issue by dluc** [PostgresApp/PostgresApp#744]
Please allow to extend the list of client apps, or at least add more options

![image](https://github.com/PostgresApp/PostgresApp/assets/371009/e7f97f28-03e6-467e-8ae0-5cfee1b75d47)

Some apps that would be great supporting:

* pgAdmin 4 - https://www.pgadmin.org
* Beekeeper Studio - https://www.beekeeperstudio.io
* DBeaver - https://dbeaver.io

many more here https://wiki.postgresql.org/wiki/PostgreSQL_Clients

Thanks!

**Comment by jakob** [comment]
It would be nice if that popup could autodetect installed PostgreSQL clients. 

I don't want to be responsible for choosing which clients are supported, and which are not.

Do all those clients support the `postgresql:` url scheme?

**Comment by dluc** [comment]
I don't think they do support the scheme. What about making the list configurable?

**Comment by tbussmann** [comment]
I once looked into the possibilities to support pgAdmin4. However, without support for an url handler nor [command line switches](https://github.com/pgadmin-org/pgadmin4/issues/4791), it seems impossible to tell the client application what database to connect to.

**Comment by jakob** [comment]
I checked the docs, and I found a pretty nice API introduced in macOS 12 that makes it easy to find apps that can handle a specific URL scheme:

```swift
import Cocoa
let appUrls = NSWorkspace.shared.urlsForApplications(toOpen: URL(string: "postgresql://localhost")!)
```

If PostgreSQL clients all supported the `postgresql:` URL scheme, that would then make it very easy to automatically add all installed apps to the popup menu.

(We would of course need to add some logic to deal with multiple copies of the same app)

Right now, Postico is the only app on my computer that supports `postgresql:` URLs. I think TablePlus may support the URL scheme (they are usually quick to add features to their app when I add them to Postico :), but I don't have it installed to check. DBeaver doesn't support it. I think I suggested adding an URL handler to one of the DBeaver devs a few years ago when I met them at a conference, but I never followed up on it.

Maybe we can suggest to app developers that they should add URL handlers to their apps. Without URL handlers and with no Apple Script support, it's going to be very hard to launch other apps...

**Comment by kantum** [comment]
Beekeeper seams to support PostgreSQL URLs:

![image](https://github.com/user-attachments/assets/18b11ee1-e097-4fa2-a2a9-34669c0a5d1b)

**Comment by jakob** [comment]
I've started to work on this, and just published a first test build here: https://github.com/PostgresApp/PostgresApp/releases/tag/v2.8alpha

I've tested it with Postico and Beekeeper Studio. I'm not sure if any other Postgres client apps work as well.

Please download the build and let me know what you think of this approach!

Also, consider contacting the developers of your favorite PostgreSQL clients and ask them to add support for the "postgres:" scheme!

**Comment by jgillard** [comment]
Just chipping in to say I tested this and TablePlus appears in the list 👍

**Comment by tbussmann** [comment]
![Bildschirmfoto 2024-11-06 um 12 07 29](https://github.com/user-attachments/assets/74f9234f-95bc-4fec-a11b-147696c8770b)
Little UI issue on macOS 12

**Comment by jakob** [comment]
TODO:

- [ ] Fix macOS 12 UI glitch
- [ ] Remove dialog for double click
- [ ] Import old setting (Terminal / iTerm / Postico) when no app selected
- [ ] Change "Always use this app" to "Use as default client application"

**Comment by tbussmann** [comment]
* [ ] Remove hint text in "PostgreSQL User"

**Comment by jakob** [comment]
I've finalized the feature in PR #776. It's ready to merge, but since it requires raising minimum macOS version to 10.15 we decided to postpone merging the feature until after next weeks release. Thanks to everyone for testing!

**Comment by jakob** [comment]
I've merged this feature yesterday. It will be published in the next release scheduled on February 13, 2025

**Comment by sersorrel** [comment]
It would be really handy to be able to pick other terminal apps besides Terminal.app and iTerm2 (for example, [Ghostty](https://ghostty.org/)), even if it's hidden behind "run `defaults write com.postgresapp.Postgres2 CustomTerminalApp /Applications/Ghostty.app`" or something like that.

**Comment by jakob** [comment]
@sersorrel Adding other clients app is a good idea. I don't think it should be a hidden setting though, Postgres.app should autodetect them if possible.

Currently we use AppleScript to control terminal apps, but Ghostty doesn't seem to support that. It looks like it can execute a command by asking it to open a shell script, though.

As a first attempt to automatically find terminal apps, I tried to use this code:
```
NSWorkspace.shared.urlsForApplications(toOpen: UTType(filenameExtension: "command")!)
```
However, that returned text editors as well. My next attempt was:
```
NSWorkspace.shared.urlsForApplications(toOpen: UTType("public.unix-executable")!)
```
but that includes HexFiend (a hex editor) as well.

I think we might have to hard code a list of terminal apps.

- iTerm2: com.googlecode.iterm2
- Ghostty: com.mitchellh.ghostty

I also checked Alacritty, but it doesn't seem to support opening shell scripts, so we probably can't support it.

Any other Terminal apps we should consider?

**Comment by sersorrel** [comment]
Checking the list that appears to be hardcoded in [Git Fork](https://git-fork.com/), they list Terminal.app, iTerm2, Hyper, Kitty, and Warp. I can't immediately think of any you haven't listed, fwiw.

**Comment by jakob** [comment]
> Maybe [MacTerm](https://www.macterm.net), [Warp](https://www.warp.dev) and [ZOC](https://www.emtec.com) also be candidates. I have never used these and unfortunately have no time to look at them at the moment

- MacTerm looks rather abandoned, there are no binaries and it seems you have to compile it yourself.
- Warp looks intersesting! Supports opening shell scripts, adding it to the list above!
- ZOC doesn't support opening shell scripts. It seems to support some Apple Script, but I was not able to find a way to make it execute a command.

**Comment by vitalyster** [comment]
When I try to connect with Beekeeper Studio it opened on start page but database not connected, is it known issue?

**Comment by jakob** [comment]
@vitalyster Good catch. Looks like connecting with Beekeeper Studio only works if the app is already running. This is an issue with Beekeeper Studio. I reported it here: https://github.com/beekeeper-studio/beekeeper-studio/issues/2818
